beacon_node/beacon_chain/src/state_advance_timer.rs

```diff
@@ -227,19 +227,51 @@ async fn state_advance_timer<T: BeaconChainTypes>(
 
                 // Prepare proposers so that the node can send payload attributes in the case where
                 // it decides to abandon a proposer boost re-org.
-                if let Err(e) = beacon_chain.prepare_beacon_proposer(current_slot).await {
-                    warn!(
-                        log,
-                        "Unable to prepare proposer with lookahead";
-                        "error" => ?e,
-                        "slot" => next_slot,
-                    );
-                }
+                let proposer_head = beacon_chain
+                    .prepare_beacon_proposer(current_slot)
+                    .await
+                    .unwrap_or_else(|e| {
+                        warn!(
+                            log,
+                            "Unable to prepare proposer with lookahead";
+                            "error" => ?e,
+                            "slot" => next_slot,
+                        );
+                        None
+                    });
 
                 // Use a blocking task to avoid blocking the core executor whilst waiting for locks
                 // in `ForkChoiceSignalTx`.
                 beacon_chain.task_executor.clone().spawn_blocking(
                     move || {
+                        // If we're proposing, clone the head state preemptively so that it isn't on the hot
```

Comment:
I wonder if we should have some logic for pruning states that are not consumed by block production (e.g., in the case that there's a proposer preparation but no actual block production).

The simplest logic might be something along the lines of *"if there's no proposer for the next slot then drop the `block_production_state`"*. 

Some more complex logic might check to see if the `block_production_state` that already exists is already suitable (e.g. we cloned the state in anticipation for slot `N`, but didn't propose. Then we had another proposal at `N+1`). I'm not convinced the complexity is worth it though.

Reply:
I've added some logic in e8c25d5789a9c5037567c51f8b09db535d7e45e2 to prune the cache when there are no proposers for the next slot and the state is more than 4 slots old. I think this should allow a limited amount of reuse by consecutive proposers, while also cleaning up memory quite quickly in the case block production is never triggered.